### PR TITLE
Instantiator Adjustments

### PR DESCRIPTION
**Pull request by korgon**
This pull request significantly improves the documentation and implementation of the `RecommendationInstantiator`, clarifies the behavior for grouped and legacy script block integrations, and adds robust test coverage for edge cases such as SPA navigation, DOM changes, and controller cleanup. The changes ensure that controllers are managed correctly when DOM elements are added or removed, and clarify how controllers are shared or separated based on profile configuration.

**Documentation Updates:**

* Added detailed documentation and examples for both grouped and legacy recommendation script block integrations, including a clear explanation that each profile entry creates a single controller and API request, regardless of how many DOM elements match its selector. [[1]](diffhunk://#diff-f7221898551db0f2bcda6506aa1a50abc92e782b34e2a5b9870cb4aa340471dfR37-R39) [[2]](diffhunk://#diff-7a66ce82de4e446c21e85f48769c0f70fd371d9a2120f0ece43b230ad2ef75b8L33-R105)

**Behavioral Clarifications and Implementation:**

* Refactored the controller creation logic so that:
  - For grouped blocks, a single controller is created per profile entry, and all matching elements share this controller.
  - For legacy blocks, each script tag results in its own controller.
  - Per-profile `DomTargeter` instances ensure each controller tracks its specific elements, and controllers are only created if the source script element is still connected to the DOM (important for SPA navigation). [[1]](diffhunk://#diff-891a76836d1f386ee1acd58b79f457d818ae8b770f6feff6b0691df9bc7e23f9R126) [[2]](diffhunk://#diff-891a76836d1f386ee1acd58b79f457d818ae8b770f6feff6b0691df9bc7e23f9L133-R135) [[3]](diffhunk://#diff-891a76836d1f386ee1acd58b79f457d818ae8b770f6feff6b0691df9bc7e23f9L151-R147) [[4]](diffhunk://#diff-891a76836d1f386ee1acd58b79f457d818ae8b770f6feff6b0691df9bc7e23f9L168-R164) [[5]](diffhunk://#diff-891a76836d1f386ee1acd58b79f457d818ae8b770f6feff6b0691df9bc7e23f9L186-R206)

**Testing Enhancements:**

* Expanded and improved test coverage to verify:
  - Controllers are cleaned up when their target elements are removed from the DOM.
  - Controllers persist when their target elements remain in the DOM.
  - The `profileCount` increments correctly when DOM is replaced and retargeted.
  - SPA navigation scenarios do not result in duplicate or stale controllers.
  - Grouped block selectors matching multiple elements share a single controller, while separate profile entries (even with the same tag) create separate controllers.
  - Attachments (plugins/middleware) are properly registered before controller creation. [[1]](diffhunk://#diff-38aaf1a2cfa852bb87208a3b953bf2939ca60dbdbafa8a52c94625c3b87eab12L224-R288) [[2]](diffhunk://#diff-38aaf1a2cfa852bb87208a3b953bf2939ca60dbdbafa8a52c94625c3b87eab12L835-R864) [[3]](diffhunk://#diff-38aaf1a2cfa852bb87208a3b953bf2939ca60dbdbafa8a52c94625c3b87eab12R893-R896) [[4]](diffhunk://#diff-38aaf1a2cfa852bb87208a3b953bf2939ca60dbdbafa8a52c94625c3b87eab12R995-R1210)

**Other Notable Improvements:**

* Added a new config option for beacon tracking settings in the documentation.
* Minor refactoring for clarity and maintainability, including importing `createRecommendationController` and improving code comments.

These changes collectively make the recommendation integration more robust, maintainable, and easier for both developers and users to understand.

---

**Documentation and Integration Behavior:**
- Added comprehensive documentation for grouped and legacy integration types, including clear explanations and warnings about controller sharing and API requests. [[1]](diffhunk://#diff-f7221898551db0f2bcda6506aa1a50abc92e782b34e2a5b9870cb4aa340471dfR37-R39) [[2]](diffhunk://#diff-7a66ce82de4e446c21e85f48769c0f70fd371d9a2120f0ece43b230ad2ef75b8L33-R105)
- Documented the new `config.beacon` option for beacon tracking settings.

**Controller Management and SPA Support:**
- Refactored controller creation to ensure:
  - Grouped block: one controller per profile entry, shared across all matched elements.
  - Legacy block: one controller per script tag.
  - Controllers are only created if the source script element is still connected (prevents orphaned controllers on SPA navigation). [[1]](diffhunk://#diff-891a76836d1f386ee1acd58b79f457d818ae8b770f6feff6b0691df9bc7e23f9R126) [[2]](diffhunk://#diff-891a76836d1f386ee1acd58b79f457d818ae8b770f6feff6b0691df9bc7e23f9L133-R135) [[3]](diffhunk://#diff-891a76836d1f386ee1acd58b79f457d818ae8b770f6feff6b0691df9bc7e23f9L151-R147) [[4]](diffhunk://#diff-891a76836d1f386ee1acd58b79f457d818ae8b770f6feff6b0691df9bc7e23f9L168-R164) [[5]](diffhunk://#diff-891a76836d1f386ee1acd58b79f457d818ae8b770f6feff6b0691df9bc7e23f9L186-R206)

**Testing and Edge Case Handling:**
- Added and improved tests to verify:
  - Controller cleanup when DOM elements are removed.
  - Controller persistence when elements remain.
  - Profile count incrementing on DOM replacement.
  - SPA navigation does not cause duplicate or stale controllers.
  - Shared vs. separate controllers for grouped profiles and selectors. [[1]](diffhunk://#diff-38aaf1a2cfa852bb87208a3b953bf2939ca60dbdbafa8a52c94625c3b87eab12L224-R288) [[2]](diffhunk://#diff-38aaf1a2cfa852bb87208a3b953bf2939ca60dbdbafa8a52c94625c3b87eab12L835-R864) [[3]](diffhunk://#diff-38aaf1a2cfa852bb87208a3b953bf2939ca60dbdbafa8a52c94625c3b87eab12R893-R896) [[4]](diffhunk://#diff-38aaf1a2cfa852bb87208a3b953bf2939ca60dbdbafa8a52c94625c3b87eab12R995-R1210)

**Code Quality and Maintainability:**
- Improved code clarity with better comments, variable naming, and the import of `createRecommendationController`.

**Test Setup Improvements:**
- Adjusted test setup to register attachments before controller creation for more accurate plugin/middleware testing. [[1]](diffhunk://#diff-38aaf1a2cfa852bb87208a3b953bf2939ca60dbdbafa8a52c94625c3b87eab12L835-R864) [[2]](diffhunk://#diff-38aaf1a2cfa852bb87208a3b953bf2939ca60dbdbafa8a52c94625c3b87eab12R893-R896)

### CODE REVIEW

File: packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx [hunk 1]
```diff
@@ -149,7 +142,11 @@ export class RecommendationInstantiator {
 				},
 			],
 			async (target: Target, elem: Element | undefined, originalElem: Element | undefined) => {
-				const elemContext = getContext(
+				// defer so this.targeter is assigned before the callback body runs (needed to attach to controller)
+				await Promise.resolve();
+
+				const scriptElement = (originalElem || elem) as HTMLScriptElement;
```
Comment (korgon) [inline packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx]: Can be cleaned up as we are no longer injecting.

File: packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx [hunk 2]
```diff
@@ -239,7 +240,22 @@ export class RecommendationInstantiator {
...
+					const profileAttr = scriptElement.getAttribute('profile') || '';
+					const recsContainer = document.createElement('div');
+					recsContainer.setAttribute('searchspring-recommend', profileAttr);
+					scriptElement.before(recsContainer);
+
+					const legacyContext = deepmerge(this.context, elemContext);
+
+					// create a per-element DomTargeter for the injected div (mirrors grouped block pattern)
+					const legacyTargeter = new DomTargeter(
+						[{ selector: `[searchspring-recommend="${profileAttr}"]`, name: `legacy_${profile}_${profileCount[profile || ''] || 0}` }],
```
Comment (korgon) [inline packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx:261]: Can we get some kind of guarantee that we will inject into the element created above?

Might need to modify DomTargeter to support passing an elem reference instead of a selector.

File: packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx [hunk 3]
```diff
@@ -239,7 +240,22 @@ export class RecommendationInstantiator {
...
+						async (_target: Target, targetElem: Element | undefined) => {
+							await Promise.resolve();
```
Comment (korgon) [inline packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx]: Can we look into using `this` instead of this workaround?

File: packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx [hunk 4]
```diff
@@ -256,6 +272,30 @@ export class RecommendationInstantiator {
...
 		this.uses.push(attachments);
 	}
+
+	// cleanup to ensure we release controllers no longer rendering in the DOM (memory leak prevention for SPA's)
+	public cleanupStaleControllers(): void {
+		Object.keys(this.controller).forEach((id) => {
+			const controller = this.controller[id];
+			const targeters = Object.values(controller.targeters);
+
+			// skip controllers that haven't registered their targeter yet (still initializing)
```
Comment (korgon) [inline packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx]: Look into if this is needed.

Comment (korgon) [reply packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx]: It is not.

File: packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx (hunk 3 above)
Comment (korgon) [reply packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx]: Only way is via another parameter in onTarget function... which is probably better to ensure synchronous targeting (speed).

File: packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx [hunk 5]
```diff
@@ -183,44 +180,48 @@ export class RecommendationInstantiator {
...
-									tag: target.profile.tag! || target.profile.profile!, // have to support both tag and profile due to having profile at release, but will favor tag
-								};
-								const profileContext: ContextVariables = deepmerge(this.context, defined({ globals: scriptContextGlobals, profile: target.profile }));
-								if (elemContext.custom) {
-									profileContext.custom = elemContext.custom;
+							const profileTargeter = new DomTargeter(
+								[profileTarget],
+								async (target: ExtendedRecommendaitonProfileTarget, elem: Element | undefined, originalElem: Element | undefined) => {
+									// defer to ensure profileTargeter is assigned (constructor fires onTarget synchronously)
+									await Promise.resolve();
```
[the remark this answers is not in the document]
Comment (korgon) [reply packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx]: This is ok - as I have moved to single controller for each selector in a profile.

File: packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx (hunk 2 above)
Comment (korgon) [reply packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx:261]: Only guarantee is that this is all very synchronous now - which I think is enough.

File: packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx [hunk 6]
```diff
@@ -183,44 +179,57 @@ export class RecommendationInstantiator {
...
-
-					// build out the targets array for each profile
+					// create a per-profile DomTargeter for each profile so each controller
+					// gets its own targeter tracking its specific element
 					scriptContextProfiles.forEach((profile) => {
 						if (profile.selector) {
-							const targetObj = {
+							const profileTarget: ExtendedRecommendaitonProfileTarget = {
 								selector: profile.selector,
 								autoRetarget: true,
```
[the remark this answers is not in the document]
Comment (korgon) [reply packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx:192]: This is correct. Because the targeter is for just the selector of the profile we only want to try to locate it initially.

File: packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx [hunk 7]
```diff
@@ -239,7 +248,26 @@ export class RecommendationInstantiator {
...
+					// create a per-element DomTargeter for the injected div (mirrors grouped block pattern)
+					new DomTargeter(
+						[{ selector: `[searchspring-recommend="${profileAttr}"]`, name: `legacy_${profile}_${profileCount[profile || ''] || 0}` }],
+						async (_target: Target, targetElem: Element | undefined, _originalElem?: Element, targeter?: DomTargeter) => {
+							// skip retarget if the source script element was removed from the DOM (SPA navigation)
+							if (!scriptElement.isConnected) {
+								return;
+							}
+
+							readyTheController(this, targetElem, legacyContext, profileCount, scriptElement, profileRequestGlobals, targeter!);
```
[the remark this answers is not in the document]
Comment (korgon) [reply packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx:268]: Not necessary.

File: packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx [hunk 8]
```diff
@@ -183,44 +179,57 @@ export class RecommendationInstantiator {
...
+
+										if (!profileControllerPromise) {
+											// first element match — create the controller
+											profileControllerPromise = readyTheController(this, targetElem, profileContext, profileCount, scriptElement, profileRequestGlobals, targeter!);
+										} else {
+											// subsequent element matches — reuse the existing controller and just render
+											const controller = await profileControllerPromise;
+											if (controller && targetElem) {
+												renderController(this, controller, targetElem, scriptElement);
+											}
```
[the remark this answers is not in the document]
Comment (korgon) [reply packages/snap-preact/src/Instantiators/RecommendationInstantiator.tsx:226]: Not necessary.

**Review by chrisFrazier77 — DISMISSED** [review]

**Review by chrisFrazier77 — APPROVED** [review]
